R/extract.R

```diff
@@ -67,15 +72,22 @@ filter_chunk_end_positions <- function(starts, ends) {
   if (length_difference == 0 && all(ends > starts)) {
     return(ends)
   }
-  if (length_difference < 0) {
-    stop("Malformed file!", call. = FALSE)
-  }
 
   positions <- sort(c(starts = starts, ends = ends))
   code_start_indexes <- grep("starts", names(positions))
-  code_ends <- positions[1 + code_start_indexes]
 
-  stopifnot(all(grepl("ends", names(code_ends))))
+  code_ends <- positions[pmin(1 + code_start_indexes, length(positions))]
+
+  bad_end_indexes <- which(grepl("starts", names(code_ends), fixed = TRUE))
+  if (length(bad_end_indexes)) {
+    bad_start_positions <- positions[code_start_indexes[bad_end_indexes]]
+    # This error message is formatted like a parse error
+    stop(sprintf(
+      "<rmd>:%1$d:1: Missing chunk end for chunk (maybe starting at line %1$d).\n",
```

Comment:
Oh wow, TIL `sprintf` accepts the `%1$d` notation! Awesome.

Reply:
Yep, I find it very handy. #544 would have to deal with that BTW.

Reply:
cc @renkun-ken

Reply:
@renkun-ken not sure if it would be helpful, but FYI Matt wrote some logic in C for handling `sprintf` fmts including handling the `%<pos>$<fmt>` notation:

https://github.com/Rdatatable/data.table/blob/master/src/snprintf.c